### PR TITLE
Emily/model standardization

### LINKED ISSUE #125: At the Model Level, Standardize both *within* tasks and *across* tasks
Currently, in our model_builder, we standardize features after the entire multi-task dataframe has been assembled; that is, after we append data from multiple tasks into a single dataframe, we then normalize each column:

```
# Feature normalization - not relevant for tree based models like XGB or RF. But will cause significant differences in the case of linear models (Linear, Lasso, etc.)
        X = pd.DataFrame(StandardScaler().fit_transform(X),columns = X.columns)
```

However, in standardizing across all columns, we capture a lot less *within-task variation*, while capturing more *between-task variation* in these features.

Consider the case of speaking time. Juries, in general, speak a lot more than teams playing CSOP. So, by standardizing across all tasks, any jury will end up with a relatively high z-score; any CSOP team --- even the most talkative CSOP team of the lot --- will have a relatively low CSOP score.

One the one hand, standardizing across all columns has benefits: it helps identify these different trends across tasks (which is good for visualization, and sanity checking!).

As a result, we have a harder time identifying how being a *relatively* less talkative Jury, or a *relatively* more talkative CSOP team, might impact your performance. We lose some predictive power at the task level.

# The Proposal
Add an option for standardization to be either within- or across- tasks, so that we can test both in models. Will we get better performance if we standardize within tasks?

### PR DESCRIPTION
**Pull request by xehu**
# Basic Info
What's this pull request about? 
> Resolve #125 --- add a flag (`standardize_within`) that standardizes all variables *within* tasks, rather than across tasks.

# Documentation
- All standardization has been moved to the `create_datasets()` function. Previously, standardization was taking place before we created the test set --- but this actually meant we (incorrectly) also standardized the task features, which is corrected in this PR.
- `standardize_within` defaults to **FALSE** (this is a design decision that I elaborate on in the next 2 bullets).
- Overall, there's a mixed impact on model performance. For CSOP (the blended model), R^2 goes up from 0.1818 to 0.1945 when you standardize within (versus across). For the 4-task joint model, the R^2 goes down from 0.1926 to 0.1896. Importantly, though, a major difference seems to be that the models that use within-standardized features have more subtle grammatical features as their top features --- things like the number of auxiliary verbs --- rather than positivity and wordcount (which were previously the top features, but these had significant between-task variation that got neutralized by the normalization).
Screenshot 2023-06-21 at 11 45 53 PM

- Given the mixed impact on model performance, my weakly-held opinion is that we should default to standardizing *across* tasks (what we had before). When we standardize within tasks, we're essentially eliminating a lot of the between-task variation that makes the task features interesting; notice how, when we standardize across tasks, the task features end up becoming much more important, because they help explain some of the between-task variation.
Screenshot 2023-06-21 at 11 52 12 PM

(I could be wrong in this decision? @yashveersinghsohi feel free to weigh in!!)

# Testing
This is what the conversation dataframe looks like, when standardizing *within* each task:
Screenshot 2023-06-21 at 11 39 38 PM
And this is the same dataframe, standardizing across all tasks. Notice that `average_num_words` has huge values for the first few rows (as those are all juries).
Screenshot 2023-06-21 at 11 40 13 PM

### CODE REVIEW

**Comment by yashveersinghsohi** [comment]
I agree with your reasoning behind setting this flag to False. I think in the current implementation we might be missing the case when the user supplies one dataset only for training (and/or testing). From the image below we can see that In this case we are not standardizing the features (the columns from the raw csop have the same distribution as the one from the preprocessed csop). 
Screen Shot 2023-06-22 at 5 52 33 PM


I have added some lines of code to do this (as comments) [here](https://github.com/Watts-Lab/team-process-map/blob/8b4196d56d8b2fca3fda532261271b723865fc8b/feature_engine/model_builder.py#L112).

If the edit makes sense, feel free to uncomment and push it. Since we have set the flag to False, this change should not really affect the `model.ipynb` notebook. It is merely a sanity check for the case when we set the flag to True and have just a single dataset in training (and/or testing).

Otherwise we are good to merge this in with main!

**Comment by xehu** [comment]
Hey @yashveersinghsohi great catch! I realized with your comment that we aren't standardizing at all in the case where we have 1 dataset. That seems incorrect, so I added standardization back in.

But upon additional thought, I feel like perhaps we should standardize regardless of the value of the `standardize_within` flag, since standardizing within and standardizing across are the same (there is 1 dataset, so there's nothing to standardize 'across'...). So I uncommented the second line but not the if statement.

Does that check out with you?

**Comment by yashveersinghsohi** [comment]
Yeah that makes sense. 
In that case, we can approve this one and merge it with main.

**Review by yashveersinghsohi — APPROVED** [review]
